File: cli/joytrunk/agent/employee_config.py
"""员工合并配置：主 config + 员工 config 覆盖（与 gateway lib/employeeConfig.js 一致）。"""

from __future__ import annotations

import json
from typing import Any

from joytrunk import paths
from joytrunk.config_schema import migrate_from_legacy


def _load_main_config() -> dict:
    p = paths.get_config_path()
    if not p.exists():
        return migrate_from_legacy(None)
    try:
        return migrate_from_legacy(json.loads(p.read_text(encoding="utf-8")))
    except Exception:
        return migrate_from_legacy(None)


def _load_employee_config(employee_id: str) -> dict | None:
    emp_dir = paths.get_employee_dir(employee_id)
    p = emp_dir / "config.json"
    if not p.exists():
        return None
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else None
    except Exception:
        return None


def _deep_merge(target: dict, source: dict) -> dict:
    out = dict(target)
    for key in source:
        if source[key] is None:
            continue
        if isinstance(source[key], dict) and isinstance(out.get(key), dict):
            out[key] = _deep_merge(out[key], source[key])
        else:
            out[key] = source[key]
    return out


def get_merged_config_for_employee(employee_id: str) -> dict:
    """主 config 打底，员工 config 中仅 agents、providers 覆盖。"""
    main = _load_main_config()
    emp = _load_employee_config(employee_id)
    if not emp:
        return main
    filtered = {}
    if isinstance(emp.get("agents"), dict):
        filtered["agents"] = emp["agents"]
    if isinstance(emp.get("providers"), dict):
        filtered["providers"] = emp["providers"]
    if not filtered:
        return main
    import copy
    return _deep_merge(copy.deepcopy(main), filtered)


def has_custom_llm(config: dict, owner_id: str | None) -> bool:
    """是否已配置自有 LLM 且当前负责人匹配。"""
    if not owner_id or config.get("ownerId") != owner_id:
        return False
    custom = (config.get("providers") or {}).get("custom") or config.get("customLLM")
    if not isinstance(custom, dict):
        return False
    return bool(custom.get("apiKey") or custom.get("apiBase") or custom.get("baseUrl"))


def get_llm_params(employee_id: str, owner_id: str) -> dict[str, Any]:
    """
    返回当前员工使用的 LLM 参数：要么自有（custom），要么 JoyTrunk Router。
    员工必须使用 LLM（用户自己的或我们的），用于与大模型对话并执行返回结果。
    """
    config = get_merged_config_for_employee(employee_id)
    defaults = (config.get("agents") or {}).get("defaults") or {}
    max_tokens = defaults.get("maxTokens") or 2048
    temperature = float(defaults.get("temperature") or 0.1)
    default_model = defaults.get("model") or "gpt-3.5-turbo"

    if has_custom_llm(config, owner_id):
        custom = (config.get("providers") or {}).get("custom") or config.get("customLLM") or {}
        return {
            "source": "custom",
            "base_url": (custom.get("apiBase") or custom.get("baseUrl") or "https://api.openai.com/v1").rstrip("/"),
            "api_key": custom.get("apiKey") or "",
            "model": custom.get("model") or default_model,
            "max_tokens": max_tokens,
            "temperature": temperature,
        }
    # 使用 JoyTrunk Router（通过 gateway 代理）
    from joytrunk.api_client import get_base_url
    return {
        "source": "router",
        "gateway_base_url": get_base_url().rstrip("/"),
        "owner_id": owner_id,
        "model": default_model,
        "max_tokens": max_tokens,
        "temperature": temperature,
    }

File: cli/joytrunk/server/lib/employeeConfig.js
/**
 * 员工级 config.json：workspace/employees/<id>/config.json
 * 可覆盖主 config 的 agents.defaults（model、maxTokens、temperature）与 providers.custom（该员工专用自有 LLM）。
 * 合并规则：主配置打底，员工配置中出现的键覆盖主配置。
 */

const fs = require('fs');
const { getEmployeeConfigPath, getEmployeeDir } = require('./paths');

function loadEmployeeConfig(employeeId) {
  const p = getEmployeeConfigPath(employeeId);
  if (!fs.existsSync(p)) return null;
  try {
    const raw = JSON.parse(fs.readFileSync(p, 'utf8'));
    return raw && typeof raw === 'object' ? raw : null;
  } catch {
    return null;
  }
}

/**
 * 深合并：target 上叠加 source，仅处理 source 中存在的键；只合并对象，其余覆盖。
 */
function deepMerge(target, source) {
  if (!source || typeof source !== 'object') return target;
  const out = { ...target };
  for (const key of Object.keys(source)) {
    if (source[key] != null && typeof source[key] === 'object' && !Array.isArray(source[key])) {
      out[key] = deepMerge(out[key] || {}, source[key]);
    } else {
      out[key] = source[key];
    }
  }
  return out;
}

/**
 * 返回该员工合并后的配置（主 config + 员工 config 覆盖）
 * getMainConfig 应返回已迁移后的主配置（如 config.loadConfig()）
 */
function getMergedConfigForEmployee(employeeId, getMainConfig) {
  const main = getMainConfig && getMainConfig();
  if (!main) return null;
  const emp = loadEmployeeConfig(employeeId);
  if (!emp || typeof emp !== 'object') return main;
  const filtered = {};
  if (emp.agents && typeof emp.agents === 'object') filtered.agents = emp.agents;
  if (emp.providers && typeof emp.providers === 'object') filtered.providers = emp.providers;
  if (Object.keys(filtered).length === 0) return main;
  return deepMerge(JSON.parse(JSON.stringify(main)), filtered);
}

/**
 * 保存员工 config：只写入允许的键（agents、providers）
 */
function saveEmployeeConfig(employeeId, data) {
  if (!data || typeof data !== 'object') return;
  const dir = getEmployeeDir(employeeId);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const out = {};
  if (data.agents && typeof data.agents === 'object') out.agents = data.agents;
  if (data.providers && typeof data.providers === 'object') out.providers = data.providers;
  fs.writeFileSync(getEmployeeConfigPath(employeeId), JSON.stringify(out, null, 2), 'utf8');
}

/** 写入空员工 config（创建员工时调用） */
function ensureEmployeeConfigExists(employeeId) {
  const p = getEmployeeConfigPath(employeeId);
  if (fs.existsSync(p)) return;
  const dir = getEmployeeDir(employeeId);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(p, '{}\n', 'utf8');
}

module.exports = {
  loadEmployeeConfig,
  getMergedConfigForEmployee,
  saveEmployeeConfig,
  ensureEmployeeConfigExists,
  deepMerge,
};
